### DISCUSSION TITLE
Unexpected evaluation of false part of if statement

### DISCUSSION THREAD

**Discussion by njcstreet** [ClickHouse/ClickHouse#62656] [Q&A] [answered]
Hi,
We have a requirement to implement a function which behaves differently based on the types of the arguments passed in. We are eventually planning to enhance the application layer to bypass the need to do this entirely, but it would be useful for a quick proof of concept. As an example we want to create a function which accepts two arguments, If both the arguments are numerical, it should add them together. If either one of the arguments is a string, it should concatenate the results. Taking the example function below:

`
CREATE OR REPLACE FUNCTION add_or_concat AS (x, y) ->
if(toTypeName(x) = 'String' OR toTypeName(y) = 'String', concat(x,  y), x + y);
SELECT add_or_concat ('a', 'b')
`
I would expect this to evaluate the left hand side of the if statement, figure out that at least one of the arguments is a String, and therefore return 'ab'. However instead we receive the error:
"Illegal types String and String of arguments of function plus" which implies that the right hand side of the statement has also been evaluated." even though short_circuit_function_evaluation is set to true.
I did some digging and I can see that in https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/toTypeName.cpp isSuitableForShortCircuitArgumentsExecution is set to false, so I wonder if this is causing the right-hand side to be executed? However we get the same issue even if we replace the entire function with:

`
CREATE OR REPLACE FUNCTION add_or_concat AS (x, y) ->
  if(1 = 1, concat(x, y), x + y);
`
Can someone explain what is going on here and provide a possible alternative? Possibly this is related to https://github.com/ClickHouse/clickHouse/issues/40622

Thanks.

**Comment by UnamedRus** [comment, marked as answer]
```
CREATE OR REPLACE FUNCTION add_or_concat AS (x, y) -> if((toTypeName(x) = 'String') OR (toTypeName(y) = 'String'), concat(x, y), toString(toUInt32OrZero(toString(x)) + toUInt32OrZero(toString(y))))

Query id: 1f849ee7-d9da-4698-93e5-3acdb8fbf53f

Ok.

0 rows in set. Elapsed: 0.009 sec.


SELECT add_or_concat(materialize('a'), materialize('b')) AS res

Query id: 16d6a323-09cc-41f5-b592-a8c6d30ad9f0

┌─res─┐
│ ab  │
└─────┘
```

**Reply by njcstreet** [reply]
Thanks very much for this. Would you be able to explain why it is required to help my understanding?

**Reply by UnamedRus** [reply]
> toUInt32OrZero(toString(x)) 

Actually innermost toString not needed.

>toUInt32OrZero

Convert any string/UInt32 to UInt32, so plus function can actually work.
If this string can't be converted to number, it return 0, but we don't care

> toString(plus(...,...))

Make returned type consistent with concat function

**Reply by njcstreet** [reply]
Thanks.
